### PR TITLE
Update ILToNativeMap event to include IL version

### LINKED ISSUE #70187: ReJITID is always zero in ILToNativeMapDCStart event
### Description

I've been looking at how we implement "Goto Source (Def)" in PerfView for JIT-compiled frames when using ETW (i.e. _not_ EventPipe)

The mapping from stack frame to source goes via a number of steps. One important piece is mapping from { module, RVA } to { Assembly, Namespace, Type, Method, IL offset }
That's achieved by tracking three tables emitted by the CLR's rundown provider: Loader/ModuleDCStart, Method/DCStartVerbose and Method/ILToNativeMapDCStart.

The first two gets us the Assembly, Namespace, Type, Method part. The final table gives us a way to find the IL offset (which, along with the PDB, allows us to identify the line number in source).

If the method in question has been re-JIT-ted -- perhaps due to tiered compilation -- we get two entries in the DCStartVerbose and ILToNativeMapDCStart tables, as you'd expect. In the DCStartVerbose table, the ReJITID can be used to distinguish them. Typically, I see 0 for "QuickJitted" and 1 for "OptimizedTier1". However, in the ILToNativeMapDCStart table the ReJITID always appears to be 0. That makes it difficult to correlate reliably (you have to rely on heuristics).

It looks like a bug that the ReJITID is always 0 in the ILToNativeMapDCStart event.

### Reproduction Steps

I can share a .diagsession, if necessary.
1. Capture a trace with PerfView of a reasonably complex .NET 6 application (something that exercises tiered compilation)
2. Open the trace, view the CPU stacks and try to "Goto Source" for some of the JITted methods. Assuming it works at all, you're likely to end up at the wrong line number. The difference may be very subtle.

More directly (which is how I found it), take a look at the raw events in the trace using PerfView's Events view. Filter to "Method/DCStartVerbose" and "Method/ILToNativeMapDCStart" events. Find an interesting method in the DCStartVerbose table -- one that has been JITted twice for tiered compilation (has a ReJITID of 1). Then search for that MethodID in the ILToNativeMap table. You'll find that it has two entries, each with ReJITID of 0. They likely have completely different IL offsets (view the raw event data) even though they claim to be the same.

### Expected behavior

ReJITID is set correctly in the ILToNativeMapDCStart event

### Actual behavior

ReJITID is always zero

### Regression?

I suspect it's been that way forever -- it's just that tiered compilation only recently became widely available.

### Known Workarounds

None

### Configuration

.NET 6, Windows, x64

### Other information

_No response_

### PR DESCRIPTION
**Pull request by davmason**
Fixes #70187

Tested that go to source continues to work with Perfview and VS CPU analysis still works.

### CODE REVIEW

**Review by jakobbotsch — COMMENTED** [review]
I don't think this will be enough for the problem they are trying to solve in #70187 for the reasons I've mentioned in https://github.com/dotnet/runtime/issues/70187#issuecomment-1164365025.

**Comment by davmason** [comment]
@jakobbotsch, yes you are correct. I opened this prematurely, in a bit I will add the native code version too

**Comment by davmason** [comment]
The approach I went with was to use the native code version as the ReJITID in the event to match the other uses, then I also created a V1 event and added the IL version too in case anyone needs it in the future

**Review by hoyosjs — APPROVED** [review]
LGTM other than that.

File: src/coreclr/vm/eventtrace.cpp [hunk 1]
```diff
@@ -6890,15 +6894,16 @@ VOID ETW::MethodLog::SendMethodILToNativeMapEvent(MethodDesc * pMethodDesc, DWOR
     // choosing to fire the event
     if ((dwEventOptions & ETW::EnumerationLog::EnumerationStructs::JitMethodILToNativeMap) != 0)
     {
-        FireEtwMethodILToNativeMap(
+        FireEtwMethodILToNativeMap_V1(
```
[the remark this answers is not in the document]
Comment (davmason) [reply src/coreclr/vm/eventtrace.cpp:6897]: Yes it is technically a breaking change, I think it is worth the slight risk though. The native code version is much more interesting when looking at IL to native maps, and it's hard to imagine anyone actually takes a dependency on knowing the IL version. The only way you'll see a new IL version is profiler rejit.

[the remark this answers is not in the document]
Comment (hoyosjs) [reply src/coreclr/vm/eventtrace.cpp:6897]: I am happy with that, just probably good do document in breaking changes release notes for the release and call it good.

**Comment by noahfalk** [comment]
This new version of the code also works properly for VS and PerfView goto source I trust?

**Comment by davmason** [comment]
> This new version of the code also works properly for VS and PerfView goto source I trust?

I did some in depth testing with an app where it did a long running operation three ways:

- All operations in tier 0
- 50/50 operations in tier 0/1
- 50/50 operations in original function and a rejitted method

Perfview handles it fine for all, VS handles it fine for tiering.

For the rejit case, VS will show two different method entries with the same name for samples:
image

That doesn't bother me at all given the relative rarity of rejit.

**Pull request merged**